### ISSUE TITLE
How to set MediaElement Latency/Caching

### ISSUE THREAD

**Issue by dmacattack** [kakone/VLC.MediaElement#66]
I'm using the MediaElement to stream RTSP, which is working well except I cant seem to figure out how to set the caching/latency. I'd like the RTSP to be as real time as possible, but it seems the vlc engine has the ingrained caching (buffering) for other video types.

For example I can set low latency in gstreamer using:
`# gst-launch-1.0 rtspsrc location=rtsp://192.168.2.1:12345 latency=30 ! decodebin ! autovideosink`

and similarly in the desktop VLC media player under 
`Tools > preferences > All > Input/Codecs > Advanced (file/network/disc caching)`

How can this be done for the MediaElement ?

**Comment by dmacattack** [comment]
It appears this can be solved with the options passing into the media element according to @kakone's README.md 

> Options : a dictionary to add some options to the media (see VLC command-line help). This property must be defined before setting the source.

But I am unsure how to do this. I have attempted to add a marquee as an example, but however I attempt to add the marquee to the vlc.MediaElement it does not work. 

For example on Linux 
`$ vlc rtsp://192.168.2.1:8080/test --sub-filter=marq --marq-marquee="hello"  `

For example on Windows 
`C:\> vlc.exe rtsp://192.168.4.108:8554/insp --sub-filter=marq --marq-marquee="hello"`

But I am unaware of the syntax for adding options in C# or in the XAML
```
// XAML
<vlc:MediaElement
   x:Name="myStream"
   AreTransportControlsEnabled="False" 
   HardwareAcceleration="True"
   AutoPlay="True"
   Width="1280"
   Height="720" >
   <!-- I get errors about the dictionary when attempting to use the XAML property "Options" -->

// C#
// no errors, but my syntax might not be correct
myStream.Options.Add("--sub-filter", "marq");
```

TLDR: How Does one add command line options to the vlc.MediaElement ?

**Comment by hieuvn** [comment]
it's seem we can''t add Option, I've tried to remove prefix"--" but it's still no luck.

**Comment by dmacattack** [comment]
Is there any way to add options to the vlc engine? My example is for the simplest option which will add "VLC" text to the upper left of the video.
![vlc_marqfont](https://user-images.githubusercontent.com/20918297/42326976-a2c929f4-8038-11e8-9546-4303cac2b232.png)

How can we pass this into the vlc.MediaElement?

**Comment by RichieRogers** [comment]
I'm also trying to understand how to use the "Options" property, but I'm finding no information on how to do it.
I am presuming that as it's a dictionary I just need to create a string, object dictionary, but as above, how do I name the options?

Richie

**Comment by RichieRogers** [comment]
Actually just found another answer that partially answers my question:
For the options, you should not include the ‘--'.
But, anyway, only a small subset of VLC options is available in libVLC for UWP. I don't think it's possible to record the video for the moment (that's why the option doesn't exist in VLC for UWP).

However, how do I know which options ARE supported?

Richie

**Comment by dmacattack** [comment]
@vbfnet 
`The VLC command syntax is very tricky. If the syntax is incorrect, VLC simply ignores it.`

Can you provide an example? Everyone seems to say it works, and that its difficult, but no one has yet provided the syntax. Why leave this up to trial and error? Many of the VLC options are not obvious, so with guessing of the syntax and guessing which options ARE available is a highly error prone process.

Can you provide a coding example of options that you have successfully used? From which, I will happily share the success I have found.

**Comment by vbfnet** [comment]
@dmacattack ,

Here is the code I wrote to record a streaming uri.

`	// To enable recording , duplicate the display to a file. 
		
	 _mediaPlayerElement.Stop();
	 
        StorageFolder folder = ApplicationData.Current.TemporaryFolder;
        string tempFolderPath = folder.Path;
		
	string fileName = "recording_1.ts";
        string recordingPath = Path.Combine(tempFolderPath, recording_1.ts");
		
	_mediaPlayerElement.Options.Remove("sout");
        _mediaPlayerElement.Options.Add("sout", "#duplicate{dst=display,dst=std{access=file,mux=ts,dst=" + recordingPath + "}}");
		
	_mediaPlayerElement.Source = uri.ToString();  // your uri
        _mediaPlayerElement.Play();
		
		...
		...
		
		// To stop recording and continue playing on monitor
		
	 _mediaPlayerElement.Stop();
	 _mediaPlayerElement.Options.Remove("sout");
         _mediaPlayerElement.Options.Add("sout", "#display{dst=display}");
	 _mediaPlayerElement.Source = uri.ToString();  // your uri
        _mediaPlayerElement.Play();`


I tried playing with volume and zoom, but with no success.

It's your turn now :-)

**Comment by vbfnet** [comment]
@dmacattack ,
I am waiting for you to share your success .

Best regards

**Comment by dmacattack** [comment]
@vbfnet 

Thank you for your posting. I was pulled off onto a different project, but now I am able to test your example. Turns out the syntax for --key=value is :
`Options.Add("key", value);`

And marquees simply just don't work (maybe not enabled in the lib), which is unfortunate because it is a glaringly obvious option.
Here is my successful implementation for RTSP options.
```
// C#
myStream.Stop();
myStream.Options.Add("network-caching", 100);
// set the source after the options
myStream.Source = "rtsp://192.168.2.1:8080/test"
myStream.Play();
```
